File: Python3/Element Appearing More Than 25% In Sorted Array.py
#1
class Solution:
    def findSpecialInteger(self, arr: List[int]) -> int:
        n = len(arr) // 4
        for i in range(len(arr)):
            if arr[i] == arr[i + n]:
                return arr[i]

#2
class Solution:
    def findSpecialInteger(self, arr: List[int]) -> int:
        def findIndexOfFirst(x):
            l = 0
            r = len(arr) - 1
            while l <= r:
                mid = (l + r) // 2
                if arr[mid] < x:
                    l = mid + 1
                else:
                    r = mid - 1
            return l

        for q in range(1, 4):
            idx = int(q * len(arr) / 4)
            start = findIndexOfFirst(arr[idx])
            if arr[start] == arr[start + len(arr)//4]:  #  check the length is at least 25%
                return arr[start]

File: Golang/Element Appearing More Than 25% In Sorted Array.go
func findSpecialInteger(arr []int) int {
    n := len(arr)
    target := n / 4
    for _, i := range []int{n / 4, n / 2, 3 * n / 4} {
        v := arr[i]
        l, r := 0, i
        for l < r {
            m := (l + r) / 2
            if arr[m] == v {
                r = m
            } else {
                l = m + 1
            }
        }
        if arr[r + target] == v {
            return v
        }
    }
    return 0
}